txtorcon/socks.py

```diff
@@ -579,7 +581,90 @@ def buildProtocol(self, addr):
 
 
 class SocksError(Exception):
-    pass
+
+    def __init__(self, message, errno=None):
+        super(SocksError, self).__init__(message)
+        self.errno = errno
+
+
```

Comment:
There's a fun trick you can play with class-args that might make all this a bit cleaner-looking. If we make each subclass include its errno / message as class-args, like this:

```
class GeneralServerFailureError(SocksError):
    message = "general SOCKS server failure"
    code = 0x01
```

Then the only class that needs and "code" at all is the superclass, which can do this:

```
class SocksError(Exception):
    def __init__(self):
        super(SocksError, self).__init__(self.message)
        # "self.code" will be accessible to all
```